### PR TITLE
Filter shipping methods by category

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

- Replaces implementation of #10474 
- Address in the same way: #9629

Reviewing #10474 I found a better place to put the logic of filtering shipping methods. So I moved the implementation there and added a unit test.

#### What should we test?

Nothing should change without activating the feature toggle `match_shipping_categories`.

From the issue:

1.  Connect as admin to a shop that has an ongoing OC
2. Edit one product (admin/products/XXX/edit) to set a specific "shipping category" (for ex "frozen")
3. In the Shipping method settings (admin/shipping_methods/XXX/edit), select another category 
4. Try to checkout with this product and this shipping method
5. See that it works
6. Repeat after activating the feature toggle only for the shop owner.

From the previous PR:

Scenario 1:
- An order with a product that belongs to the shipping category: A
- distributor supports multiple shipping methods and  none of them supports category A
=> an error should be raised
Scenario 2:
- An order with two products that belong to two different shipping categories: A and B  
- The distributor has
    -  one shipping method (sm1) that supports the shipping category A, 
    -  one shipping method (sm2) that supports the shipping category B, 
=> an error should be railse

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by abdellani — APPROVED** [review]
LGTM :+1: 

Thank you

File: app/services/order_available_shipping_methods.rb [hunk 1]
```diff
@@ -14,30 +14,40 @@ def initialize(order, customer = nil)
...
-
-    shipping_methods.uniq
+    filter_by_category(tag_rules.filter(shipping_methods))
   end
 
   private
 
-  def shipping_methods_before_tag_rules_applied
+  def filter_by_category(methods)
+    required_category_ids = order.products.pluck(:shipping_category_id).to_set
```
Comment (rioug) [inline app/services/order_available_shipping_methods.rb:23]: Today I learnt about Set !

File: spec/lib/open_food_network/tag_rule_applicator_spec.rb [hunk 2]
```diff
@@ -109,30 +109,20 @@ module OpenFoodNetwork
...
 
-      context "when the subject is nil" do
-        let(:subject) { double(:subject, reject!: false) }
-
-        it "returns immediately" do
-          applicator.filter!(subject)
-          expect(subject).to_not have_received(:reject!)
-        end
+      it "handles nil" do
+        applicator.filter(nil)
```
Comment (rioug) [inline spec/lib/open_food_network/tag_rule_applicator_spec.rb:116]: I just wondering if we should also check if it returns an empty array ?

File: spec/lib/open_food_network/tag_rule_applicator_spec.rb [hunk 3]
```diff
@@ -109,30 +109,20 @@ module OpenFoodNetwork
...
-      context "when subject is empty" do
-        let(:subject) { double(:subject, reject!: false) }
-
-        it "returns immediately" do
-          applicator.filter!(subject)
-          expect(subject).to_not have_received(:reject!)
-        end
+      it "handles an empty collection" do
+        applicator.filter(OrderCycle.none)
       end
```
Comment (rioug) [inline spec/lib/open_food_network/tag_rule_applicator_spec.rb:121]: Same as above, should we check it returns an empty array ?

**Review by rioug — APPROVED** [review]
I left a couple of comment but it shouldn't block the PR.  Looks good :+1:

File: spec/lib/open_food_network/tag_rule_applicator_spec.rb (hunk 2 above)
Comment (mkllnk) [reply spec/lib/open_food_network/tag_rule_applicator_spec.rb:116]: Good point, I didn't think about that. I just adapted the old specs. It handles empty arrays as well though. So unless we discover a bug I wouldn't add that spec.

**Comment by mkllnk** [comment]
> would it be easy to move this PR behind a dedicated feature toggle?

Yes, will do.

**Comment by mkllnk** [comment]
Feature toggle added. Ready for testing again.

**Comment by filipefurtad0** [comment]
Hey @mkllnk,

This is looking good.

With the feature-toggle **disabled**
- behavior as before -> checkout is possible despite `shipping-categories` mismatches between products and shipping-methods.

With the feature-toggle **enabled**
- after fully enabling the feature -> it is only possible to checkout when there is **at least one shipping method** which has the same `shipping-category` as the product(s) in the cart (* - see NOTE below)

- after enabling the feature for the user (`owner_id` from the enterprise):
  - we get the same behavior as fully enabled, when checking out on that store.
  - when checking out on another store the feature appears disabled - i.e., behavior as before this PR.

* NOTE
For carts with items with different shipping-categories, checkout is only possible ONLY if the shipping method has the required categories selected; this means, fully matching between the product categories.

Looks good to me.
Merging.

**Pull request merged**